### PR TITLE
Ignore `daemonset-controller-hash` label key in federation before comparing the federated object with its cluster equivalent.

### LINKED ISSUE #46925: Flake: [Federation] Daemonset history CRUD verification failures in e2e
This started happening today evening. I don't think I have enough expertise to debug this test this quickly. This is blocking the queue right now. @csbell @marun @perotinus could one of you please take a look?

Could this be related to PR #45924 which merged around the same time we started seeing this flake?

/sig federation

/assign @csbell @marun @perotinus 

cc @shashidharatd @janetkuo

### PR DESCRIPTION
**Pull request by madhusudancs**
Kubernetes daemonset controller writes a daemonset's hash to the object label as an optimization to avoid recomputing it every time. Adding a new label to the object that the federation is unaware of causes problems because federated controllers compare the objects in federation and their equivalents in clusters and try to reconcile them. This leads to a constant fight between the federated daemonset
controller and the cluster controllers, and they never reach a stable state.

Ideally, cluster components should not update an object's spec or metadata in a way federation cannot replicate. They can update an object's status though. Therefore, this daemonset hash should be a
field in daemonset's status, not a label in object meta. @janetkuo says that this label is only a short term solution. In the near future, they are going to replace it with revision numbers in daemonset status. We
can then rip this bandaid out.

Fixes #46925

**Release note**:
```release-note
NONE
```

/assign @csbell 

/cc @shashidharatd @marun @nikhiljindal @perotinus 

/sig federation

### CODE REVIEW

**Comment by madhusudancs** [comment]
cc @janetkuo 

@janetkuo do you have an issue or a design doc that describes the transition plan from the current `daemonset-controller-hash` label to `DaemonSet.Status.RevisionHistory` field?

**Comment by csbell** [comment]
What was the primary motivator for modifying the Spec in the first place rather than "doing the right thing"?

**Comment by csbell** [comment]
I'm not thrilled that we have to do this and I don't yet understand the rationale of modifying the Spec but the greater good here is to get federation back to blocking and prevent further PRs from breaking federation. Thanks for tracking this down @madhusudancs.
/lgtm
/priority failing-test

**Comment by csbell** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

File: federation/pkg/federatedtypes/daemonset.go [hunk 1]
```diff
@@ -72,6 +72,32 @@ func (a *DaemonSetAdapter) Copy(obj pkgruntime.Object) pkgruntime.Object {
...
+	// be a field in daemonset's status, not a label in object meta.
+	// @janetkuo says that this label is only a short term solution. In
+	// the near future, they are going to replace it with revision numbers
+	// in daemonset status. We can then rip this bandaid out.
+	//
+	// We are deleting the keys here and that should be fine since we are
+	// working on object copies. Also, propagating the deleted labels
+	// should also be fine because we don't support daemonset rolling
+	// update in federation yet.
+	dsHashKey := "daemonset-controller-hash"
```
Comment (janetkuo) [inline federation/pkg/federatedtypes/daemonset.go]: optional: you may want to use `DefaultDaemonSetUniqueLabelKey` defined in https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/types.go#L547

File: federation/pkg/federatedtypes/daemonset.go [hunk 2]
```diff
@@ -72,6 +72,32 @@ func (a *DaemonSetAdapter) Copy(obj pkgruntime.Object) pkgruntime.Object {
...
+	delete(daemonset1.ObjectMeta.Labels, dsHashKey)
+	delete(daemonset1.ObjectMeta.Labels, dsHashKey)
```
Comment (janetkuo) [inline federation/pkg/federatedtypes/daemonset.go]: daemonset2

**Comment by janetkuo** [comment]
Removing lgtm because of the comment above. Please add the tag back once it's fixed

File: federation/pkg/federatedtypes/daemonset.go (hunk 1 above)
Comment (madhusudancs) [inline federation/pkg/federatedtypes/daemonset.go]: Done.

File: federation/pkg/federatedtypes/daemonset.go (hunk 2 above)
Comment (madhusudancs) [inline federation/pkg/federatedtypes/daemonset.go]: Copy-paste fail. Good catch! Fixed.

**Pull request merged**